<a href="https://colab.research.google.com/github/khusanovich/modul_3/blob/main/3_2_6_foydaliFunctions.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

![Imgur](https://i.imgur.com/5pXzCIu.png)

# Data Science va Sun'iy Intellekt Praktikum

## 3-MODUL. MA’LUMOTLARGA ISHLOV BERISH

### 3.2 Ma'lumotlarni tozalash va tayyorlash

## Boshqa foydali funksiyalar

### No-odatiy qiymatlarni topish

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv("https://github.com/anvarnarz/praktikum_datasets/raw/main/MOCK_USER_DATA.csv", index_col=0)
df.head()

first_name   last_name                           email       gender   age
id                                                                          
1     Dorothy     Verryan               dverryan0@who.int   Non-binary     1
2     Pietrek     Scanlin          pscanlin1@yolasite.com         Male    91
3      Lilith    Sarjeant  lsarjeant2@creativecommons.org  Genderfluid  1233
4        Knox  Cunningham    kcunningham3@yellowpages.com  Genderfluid     6
5    Mohammed  Habbershon    mhabbershon4@sourceforge.net       Female    48

Yuqoridagi jadvlada odamlarning yoshi jamlangan. Keling jadval haqida qisqa ma'lumot olamiz:


In [3]:
df.describe()

age
count   100.000000
mean    145.660000
std     565.631411
min     -85.000000
25%      22.750000
50%      53.500000
75%      84.250000
max    4444.000000

Nimadur xato: `max` yosh `4444` deb ko'rsatayapti, tabiiyku bunday bo'lishi mumkin emas. `min` da ham manfiy qiymat ko'rishimiz mumkin, bu ham xato. Demak jadvalimizda noto'g'ri qiymatlar ham kirib qolgan. 

Keling, boshlanishiga katta qiymatlarni filtrlab olamiz, yaqin tarixda odam yoshi maksimum `123` dan oshmagnini bilgan holda, keling shu qiymatdan katta qatorlarni topamiz:

In [4]:
df[df.age>123]

first_name  last_name                           email       gender   age
id                                                                          
3       Lilith   Sarjeant  lsarjeant2@creativecommons.org  Genderfluid  1233
39      Godard      Miere               gmiere12@1688.com   Polygender  3434
47       Cleve  Sparshatt      csparshatt1a@economist.com  Genderfluid   333
49       Trace   Stockney          tstockney1c@flavors.me       Female   344
65      Eloise    Snoddon    esnoddon1s@elegantthemes.com  Genderfluid   332
80       Gabie     Rothon              grothon27@ehow.com  Genderqueer  4444
100      Mable    Toffano          mtoffano2r@myspace.com     Bigender   233

Qatorlarning indeksini ham ajratib olishimiz mumkin:

In [6]:
ind = df[df.age>123].index
ind

Int64Index([3, 39, 47, 49, 65, 80, 100], dtype='int64', name='id')

Manfiy qiymatlarni ham yuqoridagi usul bilan filtrlab olishimiz mumkin:

In [7]:
df[df.age<0]

first_name  last_name                           email       gender  age
id                                                                        
11     Elijah      Heine                eheinea@imdb.com     Bigender  -85
19      Luisa  Bradborne            lbradbornei@fema.gov       Female  -75
21      Levin     Worner                lwornerk@ihg.com  Genderqueer  -43
33   Ebenezer    Greeves           egreevesw@myspace.com         Male  -26
38       Hana   Langwade  hlangwade11@washingtonpost.com  Genderqueer  -42
90  Zachariah      Meins            zmeins2h@weather.com         Male  -69

Ikki filtrni jamlab yozish ham mumkin:

In [9]:
df[(df.age<0) | (df.age>123)]

first_name  last_name                           email       gender   age
id                                                                          
3       Lilith   Sarjeant  lsarjeant2@creativecommons.org  Genderfluid  1233
11      Elijah      Heine                eheinea@imdb.com     Bigender   -85
19       Luisa  Bradborne            lbradbornei@fema.gov       Female   -75
21       Levin     Worner                lwornerk@ihg.com  Genderqueer   -43
33    Ebenezer    Greeves           egreevesw@myspace.com         Male   -26
38        Hana   Langwade  hlangwade11@washingtonpost.com  Genderqueer   -42
39      Godard      Miere               gmiere12@1688.com   Polygender  3434
47       Cleve  Sparshatt      csparshatt1a@economist.com  Genderfluid   333
49       Trace   Stockney          tstockney1c@flavors.me       Female   344
65      Eloise    Snoddon    esnoddon1s@elegantthemes.com  Genderfluid   332
80       Gabie     Rothon              grothon27@ehow.com  Genderqueer  4444
90   Zachariah      Meins            zmeins2h@weather.com         Male   -69
100      Mable    Toffano          mtoffano2r@myspace.com     Bigender   233

Qatorlar bilan qanday yo'l tutish sizga bog'liq. Manfiy qiymatlarni musbatga o'tkazishimiz mumkin, buning uchun `numpy` dagi `abs` funksiyasini qo'llaymiz:


In [11]:
df['age'] = df['age'].apply(np.abs)
df['age'].head(10)

id
1        1
2       91
3     1233
4        6
5       48
6       77
7       67
8       99
9       77
10      68
Name: age, dtype: int64

Katta qiymatlarni tashlab yuboramiz:

In [13]:
idx = df[df.age>123].index
df.drop(index=idx, inplace=True)
df.describe()

age
count   93.000000
mean    52.612903
std     29.978265
min      1.000000
25%     29.000000
50%     52.000000
75%     77.000000
max    100.000000

Ko'rib turganingizdek, jadval qiymatlari normal holatga qaytdi.

Jadvlada yana bir e'tiborimni tortgan jihati, jinslar faqatgina erkak (`Male`) va ayol (`Female`) emas. 

In [14]:
df.gender.value_counts()

Male           21
Bigender       12
Female         11
Genderfluid    11
Non-binary     10
Genderqueer    10
Agender         9
Polygender      9
Name: gender, dtype: int64

Agar bu sizning keyingi hisoblaringizga ta'sir o'tkazadi desangiz, bu qiymatlarni ham filtrlab olishingiz mumkin:

In [15]:
df[df.gender.isin(['Male','Female'])]

first_name     last_name                         email  gender  age
id                                                                    
2     Pietrek       Scanlin        pscanlin1@yolasite.com    Male   91
5    Mohammed    Habbershon  mhabbershon4@sourceforge.net  Female   48
10      David      Meredyth    dmeredyth9@telegraph.co.uk  Female   68
12  Raffaello     Firminger     rfirmingerb@bloomberg.com  Female   74
17      Byram   Martynikhin     bmartynikhing@jiathis.com    Male   40
19      Luisa     Bradborne          lbradbornei@fema.gov  Female   75
22        Ray         Padly           rpadlyl@twitpic.com    Male   85
26   Ephrayim    Westmacott         ewestmacottp@ning.com    Male   77
29      Biron      Acklands        backlandss@comsenz.com    Male   13
33   Ebenezer       Greeves         egreevesw@myspace.com    Male   26
34     Brooks          Vose             bvosex@amazon.com  Female   34
36        Axe       Arundel           aarundelz@yahoo.com  Female   14
43       Bili     Swetenham        bswetenham16@europa.eu    Male   23
46     Alvira  Stollsteiner     astollsteiner19@mysql.com    Male   66
58     Brynna        Worsom              bworsom1l@wp.com    Male   66
61       Page       Merrell     pmerrell1o@slideshare.net    Male    1
62     Kennan         Visco          kvisco1p@jiathis.com  Female   94
63   Adrianne       Gommowe  agommowe1q@deliciousdays.com    Male  100
67     Bunnie        Ozanne      bozanne1u@whitehouse.gov    Male   92
70      Ambur        Aickin           aaickin1x@about.com    Male   60
72   Eleanora       Kamenar       ekamenar1z@examiner.com    Male   47
74      Curry    Smalecombe  csmalecombe21@washington.edu    Male   69
76      Misty         Blanc         mblanc23@amazon.co.jp    Male   74
77     Rudolf        Kalewe           rkalewe24@fotki.com    Male   12
81       Alex         Darth    adarth28@ezinearticles.com  Female   98
86    Rogerio          Akid           rakid2d@latimes.com  Female    3
87      Peder        Murrie       pmurrie2e@bloglovin.com    Male   69
88       Kain     Dumphries  kdumphries2f@hugedomains.com    Male   93
90  Zachariah         Meins          zmeins2h@weather.com    Male   69
91      Judas        Maffin         jmaffin2i@dropbox.com    Male    3
94     Sheree        Cockin     scockin2l@accuweather.com  Female   94
97     Kattie       Claxson        kclaxson2o@booking.com  Female   18

### Jadvlani aralashtirib tashlash

Ba'zida, tadqiqod uchun jadvaldagi qiymatlar o'rnini aralashtirib tashlash talab qilinishi mumkin. Misol uchun, sizda ma'lumotlar ma'lum bir tartibda turibdi, lekin siz ma'lumotlarni tasodifiy tartibda chiqarmoqchiisiz, buning uchun `numpy` kutubxonasining `random` modulidagi `permutation` funksiyasiga murojaat qilamiz.

Bizda quyidagi jadval bor:

In [16]:
df.head() 

first_name   last_name                         email       gender  age
id                                                                       
1     Dorothy     Verryan             dverryan0@who.int   Non-binary    1
2     Pietrek     Scanlin        pscanlin1@yolasite.com         Male   91
4        Knox  Cunningham  kcunningham3@yellowpages.com  Genderfluid    6
5    Mohammed  Habbershon  mhabbershon4@sourceforge.net       Female   48
6       Glynn      Ebbing        gebbing5@rakuten.co.jp  Genderfluid   77

In [17]:
len(df)

93

Jadvalimizda 93 ta qator mavjud, biz 1-93 gacha sonlarni tasodifiy tartibda saqlovchi ro'yxat yaratamiz:

In [19]:
randlist = np.random.permutation(93)

In [20]:
randlist

array([52, 32, 44,  5,  6, 40, 51, 43, 14, 64, 33, 78, 73, 84, 20, 36, 62,
       63, 41, 79, 13, 87, 89,  1, 80, 91, 92, 72, 48, 22, 69, 46, 21,  2,
       29, 34, 82, 70, 45, 35, 31, 59, 47,  4, 65, 50, 30, 90, 26, 86, 19,
       56, 10, 76, 11, 61, 17,  3, 66, 71, 83, 24, 28, 38,  8, 54, 67, 49,
       88, 55, 77,  7, 74, 81, 42, 53, 57,  0, 18, 75, 58, 60, 27, 25, 37,
        9, 16, 15, 39, 85, 23, 68, 12])

Navbat, `df` ni yuqoridagi tasodifiy tartibda tahlash, buning uchun `.take()` meotidga murojaat qilamiz

In [21]:
df.take(randlist)

first_name   last_name                         email       gender  age
id                                                                       
57   Salvador     Dusting         sdusting1k@tripod.com   Non-binary   96
34     Brooks        Vose             bvosex@amazon.com       Female   34
48  Sutherlan      Duigan             sduigan1b@ask.com   Non-binary   27
7       Wendy     Popland              wpopland6@cbc.ca  Genderfluid   67
8         Nye   Anthiftle      nanthiftle7@examiner.com      Agender   99
..        ...         ...                           ...          ...  ...
42     Dermot      Dragge        ddragge15@google.co.jp  Genderfluid   37
92       Yuri       Fahey              yfahey2j@pbs.org   Non-binary   37
25    Phaedra      Guslon      pguslono@studiopress.com  Genderqueer   18
74      Curry  Smalecombe  csmalecombe21@washington.edu         Male   69
14    Mariana     Shevels             mshevelsd@msn.com      Agender   60

[93 rows x 5 columns]

Marhamat, jadvalimiz tasodifiy tartibda tahlandi.

### Jadvaldan  tasodifiy qiymatlarni chiqarib olish.
Buning uchun bizga avvaldan tanish .sample() metodiga murojaat qilamiz va aynan nechta qiymat kerak ekanini ham ko'rsatamiz.

Bunda tasodifiy qatorlar **takrorlanmagan** holda chiqadi.

In [22]:
df.sample(10)

first_name   last_name                    email       gender  age
id                                                                  
37     Daisie        Self      dself10@addthis.com      Agender   33
16       Anny      Rodway    arodwayf@examiner.com   Non-binary   87
48  Sutherlan      Duigan        sduigan1b@ask.com   Non-binary   27
78      Richy  Cornilleau     rcornilleau25@nhs.uk     Bigender   88
6       Glynn      Ebbing   gebbing5@rakuten.co.jp  Genderfluid   77
35    Deeanne    Iamittii     diamittiiy@nifty.com     Bigender   39
87      Peder      Murrie  pmurrie2e@bloglovin.com         Male   69
68      Letty     Liddell       lliddell1v@nps.gov  Genderfluid   85
66     Alejoa  Abrahamoff    aabrahamoff1t@com.com   Polygender   28
93    Fairfax      Juleff       fjuleff2k@narod.ru   Polygender   84

In [24]:
#Agar qatorlar takrorlanishini istasangiz, `replace=True` parametrini ko'rsatasiz:
df.sample(10, replace=True)

first_name last_name                       email       gender  age
id                                                                   
81       Alex     Darth  adarth28@ezinearticles.com       Female   98
68      Letty   Liddell          lliddell1v@nps.gov  Genderfluid   85
7       Wendy   Popland            wpopland6@cbc.ca  Genderfluid   67
13    Kristan   Abriani          kabrianic@phoca.cz   Polygender   29
57   Salvador   Dusting       sdusting1k@tripod.com   Non-binary   96
71       Zane  Brickham  zbrickham1y@soundcloud.com  Genderqueer    5
6       Glynn    Ebbing      gebbing5@rakuten.co.jp  Genderfluid   77
89     Ebonee  Eastbury      eeastbury2g@smh.com.au  Genderqueer   16
97     Kattie   Claxson      kclaxson2o@booking.com       Female   18
11     Elijah     Heine            eheinea@imdb.com     Bigender   85